### DISCUSSION TITLE
AlgoMart Context Diagram

### DISCUSSION THREAD

**Discussion by houssam-hamdan** [deptagency/algomart#463] [Ideas]
Built this context diagram following the C4 model https://c4model.com that could help getting a higher picture.

![algomart_context](https://user-images.githubusercontent.com/478558/167244774-b3ef637d-02ac-4a19-bb43-abeee98f1f37.png)

@smonn I can build the following container diagram if you find this useful.

**Comment by houssam-hamdan** [comment]
@smonn I am getting to know the project, learn its stack (.net guy ;) The diagram is a draft copy and sure needs your feedback.

**Reply by smonn** [reply]
Hey! This is great, thank you for putting this together. We do have some internal diagrams like this, but not sure if I can make them public yet. Your version is pretty close to ours though. We have it split up a bit more for the central "AlgoMart system" piece and also representations of the DBs.

As the main README suggests, AlgoMart consists of four pieces of software:

1. CMS (Directus)
2. Scribe (runs various background tasks, syncs/caches data from the CMS)
3. API (using Fastify, connects CMS data with user data and blockchain data)
4. Web app (Next.js, calls the API and also connects with Firebase)

The CMS has its own database (Postgres) while Scribe and API share a database (also Postgres). We also have Redis in the mix, and we're moving towards switching over the background tasks to use BullMQ to trigger jobs in a more event-based manner.

We're a bit at capacity to provide an updated view of this, but I'll keep this in mind and circle back to you when things slow down a bit.

**Reply by houssam-hamdan** [reply]
**Scalability**
Having Redis in place acting both as distributed cache and as message broker will be great to support scalability. Syncing the cms data with the API database in an event driven style is a huge enhancement. Much better than relying on background tasks that usually comes with stability issues.

**At Capacity**
Is the secondary marketplace module still on the roadmap to be released Q2 ? This is something am waiting for ;)

**Comment by houssam-hamdan** [comment]
![algomart_container](https://user-images.githubusercontent.com/478558/170066587-aa3323f8-58ee-4b42-876e-7a790d5d5d4f.png)
Here is the drill down container diagram that visualize the different components the system is comprised of. Could you please review? 
Not sure if the background tasks communicate directly with the dbs.
@smonn
